bazaarvoice/emodb
# Added "~lastMutateAt" intrinsic

## Github Issue #

[31](https://github.com/bazaarvoice/emodb/issues/31)

## What Are We Doing Here?

This PR adds the `~lastMutateAt` intrinsic as described in the GitHub issue.  No additional features are included, although there is a slight modification to compactions.  The existing "lastMutationId" calculation in resolution and compaction is databus-centric and considers a change to the document content or tags as a mutation.  This works well to optimistically keep redundant deltas off of the databus while ensuring any potentially mutative delta is posted.  However, it interferes with the intrinsic since the `~lastMutateAt` timestamp should not be incremented if only the metadata (event tags) are changed in a delta.  The compaction record has a new field, `lastContentMutation` to track the UUID of the last delta which changed the content independently of the existing `lastMutation` field which includes tags in this calculation.  Existing compactions are grandfathered-in by substituting the `lastMutation` for the `lastContentMutation` if the former exists but the latter does not.

## How to Test and Verify

1. Check out this PR
2. Create a table called "mutation:demo"
3. Repeat this sequence:

   ```
$ curl -s -XPUT -H "Content-Type: application/json" "http://localhost:8080/sor/1/mutation:demo/demo1?audit=comment:'initial+submission'" --data-binary '{"value": true}'
{"success":true}
$ curl localhost:8080/sor/1/mutation:demo/demo1 | jq .
{
  "~lastMutateAt": "2016-09-12T22:07:21.438Z",
  "value": true,
  "~id": "demo1",
  "~table": "mutation:demo",
  "~version": 1,
  "~signature": "2cd08e5eaf149e03240259407df57c9d",
  "~deleted": false,
  "~firstUpdateAt": "2016-09-12T22:07:21.438Z",
  "~lastUpdateAt": "2016-09-12T22:07:21.438Z"
}
$ curl -s -H "Content-Type: application/x.json-delta" "http://localhost:8080/sor/1/mutation:demo/demo1?audit=comment:'unchanged'" --data-binary '{..,"value":true}'
{"success":true}
$ curl localhost:8080/sor/1/mutation:demo/demo1 | jq .
{
  "~lastMutateAt": "2016-09-12T22:07:21.438Z",
  "value": true,
  "~id": "demo1",
  "~table": "mutation:demo",
  "~version": 2,
  "~signature": "b7a9efc0bf9c8f3c07eb1b0c1518bc2a",
  "~deleted": false,
  "~firstUpdateAt": "2016-09-12T22:07:21.438Z",
  "~lastUpdateAt": "2016-09-12T22:07:41.113Z"
}
$ curl -s -H "Content-Type: application/x.json-delta" "http://localhost:8080/sor/1/mutation:demo/demo1?audit=comment:'changed'" --data-binary '{..,"value":false}'
{"success":true}
$ curl localhost:8080/sor/1/mutation:demo/demo1 | jq .
{
  "~lastMutateAt": "2016-09-12T22:08:23.446Z",
  "value": false,
  "~id": "demo1",
  "~table": "mutation:demo",
  "~version": 4,
  "~signature": "6eeb0406adfd7ba5f764bdf9529d458d",
  "~deleted": false,
  "~firstUpdateAt": "2016-09-12T22:07:21.438Z",
  "~lastUpdateAt": "2016-09-12T22:08:23.446Z"
}
$ curl -s -H "Content-Type: application/x.json-delta" "http://localhost:8080/sor/1/mutation:demo/demo1?audit=comment:'unchanged'" --data-binary '{..,"value":false}'
{"success":true}
$ curl localhost:8080/sor/1/mutation:demo/demo1 | jq .
{
  "~lastMutateAt": "2016-09-12T22:08:23.446Z",
  "value": false,
  "~id": "demo1",
  "~table": "mutation:demo",
  "~version": 5,
  "~signature": "4c8d7853afae28285cf7bba5fece5dbd",
  "~deleted": false,
  "~firstUpdateAt": "2016-09-12T22:07:21.438Z",
  "~lastUpdateAt": "2016-09-12T22:08:33.204Z"
}
   ```

## Risk

Medium.  In addition to the risk outlined in the issue, the added attribute to compaction adds some risk, particularly with regards to handling of existing compaction records.

### Level 

Medium

### Required Testing

Manual.

### Risk Summary

No additional risks not previously discussed.

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
